### ISSUE TITLE
[torch.compile] Integers stored on nn.Modules as dynamic causing errors

### ISSUE THREAD

**Issue by imzhuhl** [pytorch/pytorch#133166]
### 🐛 Describe the bug

reporduce code
```python
from transformers import BloomModel, BloomConfig
import json
import torch

CONFIG = """
{
  "apply_residual_connection_post_layernorm": false,
  "architectures": [
    "BloomModel"
  ],
  "attention_dropout": 0.0,
  "attention_softmax_in_fp32": true,
  "bias_dropout_fusion": true,
  "bos_token_id": 1,
  "eos_token_id": 2,
  "hidden_dropout": 0.0,
  "hidden_size": 1536,
  "initializer_range": 0.02,
  "layer_norm_epsilon": 1e-05,
  "masked_softmax_fusion": true,
  "model_type": "bloom",
  "n_head": 16,
  "n_inner": null,
  "n_layer": 24,
  "offset_alibi": 100,
  "pad_token_id": 3,
  "pretraining_tp": 1,
  "seq_length": 2048,
  "skip_bias_add": true,
  "skip_bias_add_qkv": false,
  "slow_but_exact": false,
  "torch_dtype": "float32",
  "transformers_version": "4.29.2",
  "unk_token_id": 0,
  "use_cache": false,
  "vocab_size": 250880
}
"""

config = BloomConfig(**json.loads(CONFIG))
config.n_layer = 1
model = BloomModel(config).half().cuda().eval()

x = torch.randint(100, config.vocab_size, (1, 20)).cuda()

with torch.no_grad():
    ref_out = model(x)[0]

    compiled_fn = torch.compile(model, dynamic=True)
    out = compiled_fn(x)[0]
```
The above code is fine on 2.3, but it will throw an error on 2.4 and the main branch.
What is error looks like:
```
def triton_(in_ptr0, in_ptr1, out_ptr0, ks0, ks1, ks2, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x1 = (xindex // ks2)
    x0 = xindex % ks2
    x2 = xindex
    tmp5 = tl.load(in_ptr0 + (x0), xmask, eviction_policy='evict_last')
    tmp8 = tl.load(in_ptr1 + (x0), xmask, eviction_policy='evict_last')
    tmp0 = (ks1.to(tl.float64)) ** ((-1)*((ks0.to(tl.float64)) ** (-1.00000000000000)))
                                          ^
AttributeError("'tensor' object has no attribute '__pow__'")

Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
```
This seems to be a problem with inductor or triton, but it is actually because *self.num_heads* is dynamic.
This phenomenon occurred after this commit: https://github.com/pytorch/pytorch/issues/115711

The problem is that torch made all ints in nn.Module dynamic, but this is actually not necessary. I understand that there will be an automatic dynamic mechanism for this in the future, but for the moment can I have a simple way to avoid the problems.

### Versions

```
PyTorch version: 2.4.0+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Alibaba Cloud Linux release 3 (Soaring Falcon)  (x86_64)
GCC version: (GCC) 10.2.1 20200825 (Alibaba 10.2.1-3.6 2.32)
Clang version: Could not collect
CMake version: version 3.26.5
Libc version: glibc-2.32

Python version: 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.10.134-16.1.al8.x86_64-x86_64-with-glibc2.32
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA A10
Nvidia driver version: 550.54.15
cuDNN version: Probably one of the following:
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn.so.8.9.7
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.9.7
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.9.7
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.9.7
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.9.7
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.9.7
/usr/local/cuda-12.4/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.9.7
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
架构：           x86_64
CPU 运行模式：   32-bit, 64-bit
字节序：         Little Endian
CPU:             48
在线 CPU 列表：  0-47
每个核的线程数： 2
每个座的核数：   24
座：             1
NUMA 节点：      1
厂商 ID：        GenuineIntel
BIOS Vendor ID:  Alibaba Cloud
CPU 系列：       6
型号：           106
型号名称：       Intel(R) Xeon(R) Platinum 8369B CPU @ 2.90GHz
BIOS Model name: pc-i440fx-2.1
步进：           6
CPU MHz：        2899.998
BogoMIPS：       5799.99
超管理器厂商：   KVM
虚拟化类型：     完全
L1d 缓存：       48K
L1i 缓存：       32K
L2 缓存：        1280K
L3 缓存：        49152K
NUMA 节点0 CPU： 0-47
标记：           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq monitor ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves wbnoinvd arat avx512vbmi avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid fsrm arch_capabilities

```

**Comment by imzhuhl** [comment]
pytorch 2.4 release error infomation:
![image](https://github.com/user-attachments/assets/7279f7cc-2dde-4705-bb28-20dcad7c64ee)

**Comment by shunting314** [comment]
cc @yanboliang who authored: https://github.com/pytorch/pytorch/issues/115711

**Comment by ezyang** [comment]
Adding insult to injury there's no way to force the integer to be static lol. Well, I guess you could do something like `tensor_num_heads = torch.empty((num_heads, 0))` and then `num_heads = tensor_num_heads.size(0)` LOL

**Comment by ezyang** [comment]
It looks like this is causing many errors internally

**Comment by imzhuhl** [comment]
I actually circumvent this problem by using warmup. I don't specifiy dynamic=True, but use default setting. Then I warmup with inputs of different lengths. Eventually, the dimension of the sequence length become dynamic, while num_heads is still static, which serves my purpose.

**Comment by sbucaille** [comment]
Hi, I originally mentionned this issue from my PR [here](https://github.com/huggingface/transformers/pull/31718). I'm not very familiar with torch.compile and how it works internally but I found a workaround, I don't know if it is of any help for you but here it is : 

Original code
```python
class LightGlueMatchAssignmentLayer(nn.Module):
    def __init__(self, config: LightGlueConfig):
        super().__init__()

        self.descriptor_dim = config.descriptor_dim

    def forward(self, descriptors: torch.Tensor, ...) -> torch.Tensor:
        # ...
        m_descriptors = m_descriptors / self.descriptor_dim**0.25
```

Workaround
```python
class LightGlueMatchAssignmentLayer(nn.Module):
    def __init__(self, config: LightGlueConfig):
        super().__init__()

        self.descriptor_dim = config.descriptor_dim

    def forward(self, descriptors: torch.Tensor, ...) -> torch.Tensor:
        # ...
        m_descriptors = m_descriptors / torch.pow(torch.tensor(self.descriptor_dim), 0.25)
```

**Comment by anijain2305** [comment]
@shunting314  @eellison I am getting this error right now

```
def triton_poi_fused__to_copy_1(out_ptr0, ks0, ks1, xnumel, XBLOCK : tl.constexpr):
    xoffset = tl.program_id(0) * XBLOCK
    xindex = xoffset + tl.arange(0, XBLOCK)[:]
    xmask = xindex < xnumel
    x1 = xindex // ks1
    x0 = (xindex % ks1)
    x2 = xindex
    tmp0 = libdevice.pow(tl.full([], 2.00000000000000, tl.float64), (-1)*libdevice.pow(tl.full([], 2.00000000000000, tl.float64), tl.full([], 3.00000000000000, tl.float64) + ((-1)*OpaqueUnaryFn_log2((libdevice.pow(2, libdevice.floor(OpaqueUnaryFn_log2(ks0.to(tl.float64))).to(tl.int32))).to(tl.float64)))))
                                                                                                                                                                                    ^
NameError('OpaqueUnaryFn_log2 is not defined')
```

Any way to circumvent this error?

**Comment by shunting314** [comment]
How are those sympy functions like OpaqueUnaryFn_log2  suppose to be lowered? cc @ezyang

**Comment by isuruf** [comment]
https://github.com/pytorch/pytorch/pull/147592 should fix this, but I could use some help in writing a test-case.

**Comment by isuruf** [comment]
Is it okay to use the original issue with BloomModel as a test case?

**Comment by isuruf** [comment]
I used a smaller test-case in #147592

**Comment by anijain2305** [comment]
The test case works now. But an action item on me is to look closer into the dynamism of nn module attributes.
